src/mono/wasm/Wasm.Build.Tests/TestAppScenarios/SignalRClientTests.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.Playwright;
+using Xunit.Abstractions;
+using Xunit;
+
+#nullable enable
+
+namespace Wasm.Build.Tests.TestAppScenarios;
+
+public class SignalRClientTests : AppTestBase
+{
+    public SignalRClientTests(ITestOutputHelper output, SharedBuildPerTestClassFixture buildContext)
+        : base(output, buildContext)
+    {
+    }
+
+    [ConditionalTheory(typeof(BuildTestBase), nameof(IsWorkloadWithMultiThreadingForDefaultFramework))]
+    [InlineData("Debug", "LongPolling")]
+    [InlineData("Release", "LongPolling")]
+    [InlineData("Debug", "WebSockets")]
+    [InlineData("Release", "WebSockets")]
+    public async Task SignalRPassMessages(string config, string transport)
+    {
+        CopyTestAsset("BlazorHostedApp", "SignalRClientTests");
+        string frameworkDir = FindBlazorHostedBinFrameworkDir(config,
+            forPublish: false,
+            clientDirRelativeToProjectDir: "../BlazorHosted.Client");
+        BuildProject(configuration: config,
+            binFrameworkDir: frameworkDir,
+            runtimeType: RuntimeVariant.MultiThreaded);
+
+        List<string> testOutput = new();
+        List<string> consoleOutput = new();
+        List<string> serverOutput = new();
+
+        // To look for static files in the paths defined by "*staticwebassets.runtime.json"
+        // and avoid: "Static files may be unavailable."
+        s_buildEnv.EnvVars["ASPNETCORE_ENVIRONMENT"] = "Development";
+        using var runCommand = new RunCommand(s_buildEnv, _testOutput)
```

Comment:
You passing in `_testOutput` instead of `testOutput`, is it intended?
(the same one line below)

Reply:
The idea was to keep 2 of them for different purpose: `testOutput` tracks the actions connected with signalR and the flow of this specific scenario (hub connected / message sent etc),  inherited `_testOutput` tracks the "behind the scenes" actions, like: browser got opened, page logged error etc. But in case of exception, only `_testOutput` would be shown and this is not the best approach, we can merge them and leave `_testOutput` only.

Reply:
Coping the test app and building is fine. I would like to unify the code needed to start the app, connect browser and waiting for load. Because that is common for all tests.

The `RunSdkStyleAppForBuild` is currently missing callbacks for making the test interactive, but the rest of the logic is there. My idea about setup is

```csharp
var result = await RunSdkStyleAppForBuild(new(
    Configuration: config,
    ServerEnvironment: new Dictionary<string, string> { ["ASPNETCORE_ENVIRONMENT"] = "Development" },
    TestScenario: "",
    BrowserPath: "/chat",
    BrowserQueryString: ...,
    OnPageLoaded: async page =>
    {
        await page.ClickAsync("button#connectButton");
        // give the connection some time to establish
        await Task.Delay(3000);
    },
    OnConsoleMessage: async (msg, page) =>
    {
        if (msg.Text.Contains("SignalR connected"))
            await page.ClickAsync("button#subscribeButton");

        ...
    }
));
```

The `OnPageLoaded` is what `BlazorRunOptions.Test` is under the hood. The complicated thing is to pass `page` to `OnConsoleMessage`.